From 958b495636dc409bd9f7803d24e86a3902541a97 Mon Sep 17 00:00:00 2001
From: Stanley Tng <stng@google.com>
Date: Wed, 28 Mar 2018 17:12:28 -0700
Subject: [PATCH] DO NOT MERGE Drop LE CoC fragments when frame size is too big

Drop the LE CoC data fragments when the received fragment size is too
big.

Test: Runs LE CoC SL4A test, BleCocTest.
Bug: 75298652
Merged-In: I529944341e9e67a39e7ec7e740d5ada3db8cc23a
Change-Id: I529944341e9e67a39e7ec7e740d5ada3db8cc23a
(cherry picked from commit 8365a2ace5e89d8b81bab468f0f9bc1137d773b4)
(cherry picked from commit 17db92e4fc3c7127c0ace625ff9735a9972eee70)
CVE-2018-9380
---

diff --git a/stack/l2cap/l2c_fcr.c b/stack/l2cap/l2c_fcr.c
index 33e2a41..6b7586c 100644
--- a/stack/l2cap/l2c_fcr.c
+++ b/stack/l2cap/l2c_fcr.c
@@ -24,6 +24,7 @@
  ******************************************************************************/
 
 #include <assert.h>
+#include <log/log.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -896,7 +897,24 @@
 
     }
     else
+    {
         p_data = p_ccb->ble_sdu;
+        if (p_buf->len > (p_ccb->ble_sdu_length - p_data->len)) {
+          L2CAP_TRACE_ERROR("%s: buffer length=%d too big. max=%d. Dropped",
+                            __func__, p_data->len,
+                            (p_ccb->ble_sdu_length - p_data->len));
+          android_errorWriteWithInfoLog(0x534e4554, "75298652", -1, NULL, 0);
+          osi_free(p_buf);
+
+          /* Throw away all pending fragments and disconnects */
+          p_ccb->is_first_seg = true;
+          osi_free(p_ccb->ble_sdu);
+          p_ccb->ble_sdu = NULL;
+          p_ccb->ble_sdu_length = 0;
+          l2cu_disconnect_chnl(p_ccb);
+          return;
+          }
+    }
 
     memcpy((UINT8*)(p_data + 1) + p_data->offset + p_data->len, (UINT8*)(p_buf + 1) + p_buf->offset, p_buf->len);
     p_data->len += p_buf->len;
@@ -912,11 +930,6 @@
     {
         p_ccb->is_first_seg = FALSE;
     }
-    else
-    {
-        L2CAP_TRACE_ERROR ("%s Length in the SDU messed up",__func__);
-        // TODO: reset every thing may be???
-    }
 
     osi_free(p_buf);
     return;
